src/snapshot.ts

```diff
@@ -16,11 +16,21 @@ export function getSnapshot<T extends IAnyType>(value: IStateTreeNode<T>): Snaps
 
 const snapshot = (value: any): unknown => {
   if (value instanceof QuickArray) {
-    return Array.from(value.map((v) => snapshot(v)));
+    const type = getType(value) as ArrayType<IAnyType>;
+    const childrenAreReferences = isReferenceType(type.childrenType);
+
+    return Array.from(value.map((v) => (childrenAreReferences ? v[$identifier] : snapshot(v))));
```

Comment:
Note that we have to do it this way because reference types are "lossy" in readonly roots, meaning that `v` here is an instance of `ModelType`, not some special kind of `ReferenceType` instance (see how we [instantiate](https://github.com/gadget-inc/mobx-quick-tree/blob/e82181a76749d6d1310865e0ab892f8d8643267d/src/reference.ts#L31-L36) refs)

Reply:
Ah, I see, I see.

Reply:
Yeah, I was hoping I could do something like `isReferenceType` at the root of the snapshot function, but alas :(

Reply:
Another option would be to put the snapshot generator functions on all the types themselves -- a reference type can know how to snapshot itself, and an array can know how to snapshot itself by calling `.map(el => this.elementType.snapshot(el)` etc, but whats here works fine too 👍

Reply:
👍 sounds like a great refactor. I'll go ahead with this to make progress on other Gadget things, but create an issue for it!